README
Kapi - A keyframing API for the HTML 5 Canvas
=============================================

What Kapi does
--------------

Kapi is a keyframing API for the HTML 5 canvas.  It streamlines animation development by providing a simple interface for defining “actors” and their changing states over time.

A critical component of programmatically creating an animation is accurately calculating where things should be at a given time.  Doing something like

#+BEGIN_SRC
function update() {
  myActor.x++;
  setTimeout(update, 33);
}
#+END_SRC

is brittle, because it will lose accuracy.  System resource availability will affect the actual state of myActor over time.  A better approach is calculating the state of something based on real time.  This is precisely what Kapi does for you.

The important thing to understand is that Kapi is simply a timing layer for your animations.  It is designed to be modular, so you can use whatever drawing layer you prefer.

How Kapi works
--------------

At a high level, using Kapi comes down to five steps:
Create a Kapi instance.
Create some actors.
Add the actors to the Kapi instance.
Define keyframes (states) for the actors.
Start the animation.


1.  A Kapi animation is comprised of two metaphorical components:
2.  A stage.
3.  Actors.


The stage is a place for actors to do what they do.  In the literal sense, the stage is an HTML 5 canvas object, and actors are JavaScript functions that draw something on the canvas.  Here is a very simple example of an actor function that is compatible with Kapi:

#+BEGIN_SRC
function circle(canvasContext){
  canvasContext.beginPath();
  canvasContext.arc(
    this.x,
    this.y,
    this.radius,
    0,
    Math.PI*2, 
    true
  );
  canvasContext.fillStyle = this.color;
  canvasContext.fill();
  canvasContext.closePath();
}
#+END_SRC

Kapi only does one thing - it manages values based on the passage of time.  It passes those values to your actors, and your actors can do whatever you want them to with the values.

Using Kapi

Setting up your animation
-------------------------

The first step is to create an instance of the Kapi object.  This can be done as simply as:

#+BEGIN_SRC
var myKapi = kapi(document.getElementsByTagName('canvas')[0]);
#+END_SRC

`myKapi` is the object by which you will interface with the canvas.  There are optional parameters and events that you can pass along to the `kapi` constructor to configure the instance - please consult the public API documentation for more info.

The next step is to add actor instances.  You have complete freedom on how to you want to define your actors.  They are simply doing stuff with their instance variables, which are accessed with the JavaScript `this` keyword.  Let’s use the “circle” example actor defined earlier.  To add an instance of “circle” to Kapi, simply use `.add()`:

#+BEGIN_SRC
var myCircle = myKapi.add(circle, {	
    name : 'myCircle',
    x : 0,
    y : 0,
    radius : 50,
    color : '#00ff00',
    easing : 'easeInOutQuad'
 });
#+END_SRC

`kapi.add()` defines the initial state of an actor - which is done in the second parameter with an object.  The properties of this object will be present in all keyframes of myCircle.  If there is a property that will be used anywhere by this actor in the animation, you need to define it in `kapi.add().`

A few important things to note:  If you do not specify a `name` property, one will be randomly generated for you.  It does need to have the same name as the variable that gets assigned the reference of the returned object from `kapi.add().`

You can add as many actors as you like with this approach.  Also, you can mix and match the actor types as you please (you can have multiple `circle`s.)

Defining states
---------------

Now that you have set up your actors and your stage, you can define keyframed states.

To define a keyframe, call “actor.keyframe()”, like so:

#+BEGIN_SRC
myCircle.keyframe(0, {
  x : 200,
  y : 150,
  radius : '*=2',
  color : '#0000ff'
});
#+END_SRC

This would make a keyframe for `myCircle` at keyframe zero.  This particular keyframe would place the circle down and to the right, double the size, and fade it to blue (from green).  Since this is keyframe zero, the change would be completed as soon as the animation begins.  Any properties that are missing from a keyframe definition are inherited from the previous keyframe, if there is one.  If the property isn’t present on that keyframe, the same property from the keyframe before that is used.  This property lookup goes all the way back to the state that was passed to `kapi.add()` if necessary.

Actors will not appear on the screen until they are given a keyframe.  If the first keyframe that an actor is on is also the last keyframe in the animation, you won’t see that actor because the animation is starting over as soon as the actor “enters” the stage.

The `actor.keyframe()` function is chainable, so you can do this:

#+BEGIN_SRC
myCircle.keyframe(0, {
  x : 60,
  y : 50,
  color : '#0000ff'
}).keyframe(30, {
  x : 250,
  y : 50,
  radius : '/=2'
});
#+END_SRC

Immediate actions


Another way to move your actors around is to use an Immediate Action.  Currently the only Immediate Action is `actor.to().`  Immediate actions work a little differently that keyframes.  Keyframes, collectively, form an animation loop that will repeat itself.  Immediate Actions, on the other hand, are executed immediately, and are executed only once and then discarded.  If multiple Immediate Actions are created for an actor, they are placed into a queue and fire in the order that they were created.

#+BEGIN_SRC
myCircle.to('2s', {
  x: '+=100',
  y: 50,
  color: '#ff0000'
});
#+END_SRC

From the instant this method is called, `myCircle` will move 100 pixels to the right of its current position, to the 50th pixel down in the canvas, and fade to red.  This will happen over the course of two seconds, which is defined by the first parameter.

Immediate Actions, like keyframes, are chainable.  It is easiest to use Immediate Actions with animations that only have one keyframe, defined on keyframe zero.  However, mixing Immediate Actions with keyframe is entirely valid.

#+BEGIN_SRC
myCircle.keyframe(0, {
  x : 100,
  y : 100,
  color : '#0000ff'
}).to('2s', {
  x: '+=100',
  y: 50,
  color: '#ff0000'
}).to('1s', {
  x: 75,
  y: ‘*=2’
});
#+END_SRC

This snippet defines keyframe zero, and adds two Immediate Actions to the queue.  In total, this animation will run for 3 seconds and then just stop.

Controlling Kapi
----------------

Once you have set up your stage and actors, you can start and stop the animation.  Additionally, you can specify any point in the animation’s timeline that you’d like to view and play from.  Using our `myKapi` instance from before, we can call the following methods:

#+BEGIN_SRC
myKapi.play();
#+END_SRC

Runs the animation from the beginning if it was not running before, or resumes from the paused state.

#+BEGIN_SRC
myKapi.pause();
#+END_SRC

Pauses the animation, but does not clear the canvas.  The state of the animation is “frozen” until it is started again.

#+BEGIN_SRC
myKapi.stop();
#+END_SRC

Stops the animation, clears, the canvas, and resets the state back to the beginning of the animation loop.

If you would like to go to a specific point in the timeline of the animation, you can do so with 

#+BEGIN_SRC
myKapi.gotoFrame(desiredFrame);
#+END_SRC

desiredFrame is any frame in the animation.  For your convenience, there is also 

#+BEGIN_SRC
myKapi.gotoAndPlay(desiredFrame);
#+END_SRC

This simply calls  “gotoFrame()” and then “play()”.

Tweening
--------

There is only one tweening method built into Kapi - Linear.  Tweens are extensible; just add methods to the global `kapi.tween` object.  Please consult the docs for more info regarding this.

If you want more tweens than Linear, you are in luck.  The file `shapes.js` is included in the Kapi repository, which contains a collection of tweening methods.  All you have to do is include `shapes.js` in your page to use all the fun tweens.

More info
---------

More technical documentation can be found within the `kapi.js` source file itself.  Each method has an explanation of the expected parameters and output.

changelog.md
Changelog
===

* __1.0.7__: 2011-05-06.  Added parameters to `kapi.add()` and the actor template's `setup` method. SHA: 75c9ee57945d52a45f715a9c9a75f839713261ca

* __1.0.6__: 2011-05-03.  Fixed [Issue 55](https://github.com/jeremyckahn/kapi/issues/55): "kapi.framerate() might mess up statically timed keyframes".  SHA: ce3783570511af21d4b410b060701119462db2bf

* __1.0.5__: 2011-04-28.  Added [`actor.data()`](http://jeremyckahn.github.com/kapi/actor_doc.html#data).  Actor `draw()` functions now receive the actor object as a parameter (this means a draw function can access actor methods).  SHA: b708575e5380a91a056e2a9f701a8e6e7805f9fc

* __1.0.4__: 2011-04-28.  Fixes [issue 52](https://github.com/jeremyckahn/kapi/issues/52).  Removed a global tween variable.  SHA: 4bfd452063adb89baf75a182d4e7551dda77c1bd

* __1.0.3__: 2011-04-24.  Added `.getContext().`  Actors now have reference to their `kapi` instance through `actorObj.kapi`.  SHA: a1bcdaa8c7b8c1cf0ab2573ae08fbc514f07a3a4

* __1.0.2__: 2011-04-24.  Added `.redraw().` SHA: 6a3348b8a721ea62bf1c1ef0eccd7e508cc26d9d

* __1.0.1__: 2011-04-23.  Fixed a bug in `kapi.gotoFrame()` and `kapi.gotoAndPlay()`.  Fixes [issue 41](https://github.com/jeremyckahn/kapi/issues/41).  Added `kapi.clear()`.  Added options `clearOnComplete` and `clearOnStop`, both of which are `false` by default. SHA: 8a81bc714c151ef63b41af8ad5c4df1a82f79dea

* __1.0.0__: 2011-04-09.  Initial stable release. SHA: e3b089e3b239eabfbcb040fed894e7b468b73772

js/shapes.js
/**
 * These functions have nothing to do with Kapi core code.  They are just useful for testing it out.
 * Actors can be functions or objects.
 */


var circle = {
	setup: function (kapiInst, actorInst, params) {
		if (window.console) {
			console.log('Setting up: ' + this.id)
		}
	},
	draw: function (ctx, kapiInst, actorInst){
		ctx.beginPath();
		ctx.arc(
			this.x || 0,
			this.y || 0,
			this.radius || 0,
			0,
			Math.PI*2, 
			true
			);
		ctx.fillStyle = this.color || '#f0f';
		ctx.fill();
		ctx.closePath();

		return this;
	},
	teardown: function (actorName, kapiInst) {
		if (window.console) {
			console.log('Tearing down ' + actorName);
			console.dir(kapiInst);
		}
	}
}


function square(ctx){
	ctx.beginPath();
	
	ctx.moveTo(this.x, this.y);
	ctx.lineTo(this.x + this.width, this.y);
	ctx.lineTo(this.x + this.width, this.y + this.height);
	ctx.lineTo(this.x, this.y + this.height);
	
	ctx.fillStyle = ctx.strokeStyle = this.color || '#f0f';
	ctx.fill();
	ctx.stroke();
	ctx.closePath();
}


js/demo_runner.js
$(function () {
	
	SyntaxHighlighter.all();
	
	if (!$('canvas').length) {
		return;
	}
	
	/* Demo 1 ******************************************/
	(function () {
		var myKapi,
			circle1;
		
		myKapi = kapi(document.getElementById('demo1'), {
			fps: 30,
			// Set up some basic styles for the canvas
			styles: {
				height: 	'400px',
				width: 		'500px',
				background: '#333'
			}
		}).play(); // Start the animation
		
		// The actor `circle` is defined in shapes.js
		circle1 = myKapi.add(circle, {
			name: 		'circle1',
			x: 		70,
			y: 		70,
			radius: 50,
			color: 	'#00ffaa',
			easing: 'easeInOutQuint'
		}).keyframe(0, {
			// Blank, defining keyframe zero as the same as the initial state
		}).keyframe('3s', {
			x: 	300,
			y: 	300
		}).keyframe('3.25s', {
			// Empty keyframes act as a "wait"
		}).keyframe('6s', {
			x: 	'+=50',
			y: 	100
		}).liveCopy('8s', 0);
		
		// Some jQuery code to attach a click event for pausing.
		// It is not necessary to get the above code to run.
		$('#demo1').click(function (ev) {
			if (myKapi.isPlaying()) {
				myKapi.pause();
			} else {
				myKapi.play();
			}
		});
		
		// Makes the myKapi variable global.  Not neccesary, just handy for checking kapi.min.js.
		window.myKapi = myKapi;
	}());
	/****************************************** Demo 1 */
});

js/kapi.tweens.js
// All equations are copied from here: http://www.gizma.com/easing/
// Originally written by Robert Penner, copied under BSD License (http://www.robertpenner.com/)
//
// Params are as follows
// t = current time
// b = start value
// c = change in value
// d = duration
kapi.tween.easeInQuad = function (t, b, c, d) {
	t /= d;
	return c * t * t + b;
};

// decelerating to zero velocity
kapi.tween.easeOutQuad = function (t, b, c, d) {
	t /= d;
	return -c * t * (t - 2) + b;
};

// acceleration until halfway, then deceleration
kapi.tween.easeInOutQuad = function (t, b, c, d) {
	t /= d / 2;
	if (t < 1) {
		return c / 2 * t * t + b;
	}
	t--;
	return -c / 2 * (t * (t - 2) - 1) + b;
};

// accelerating from zero velocity
kapi.tween.easeInCubic = function (t, b, c, d) {
	t /= d;
	return c * t * t * t + b;
};

// decelerating to zero velocity
kapi.tween.easeOutCubic = function (t, b, c, d) {
	t /= d;
	t--;
	return c * (t * t * t + 1) + b;
};

// acceleration until halfway, then deceleration
easeInOutCubic = function (t, b, c, d) {
	t /= d / 2;
	if (t < 1) {
		return c / 2 * t * t * t + b;
	}
	t -= 2;
	return c / 2 * (t * t * t + 2) + b;
};

// accelerating from zero velocity
kapi.tween.easeInQuart = function (t, b, c, d) {
	t /= d;
	return c * t * t * t * t + b;
};

// decelerating to zero velocity
kapi.tween.easeOutQuart = function (t, b, c, d) {
	t /= d;
	t--;
	return -c * (t * t * t * t - 1) + b;
};

// acceleration until halfway, then deceleration
kapi.tween.easeInOutQuart = function (t, b, c, d) {
	t /= d / 2;
	if (t < 1) {
		return c / 2 * t * t * t * t + b;
	}
	t -= 2;
	return -c / 2 * (t * t * t * t - 2) + b;
};

// accelerating from zero velocity
kapi.tween.easeInQuint = function (t, b, c, d) {
	t /= d;
	return c * t * t * t * t * t + b;
};

// decelerating to zero velocity
kapi.tween.easeOutQuint = function (t, b, c, d) {
	t /= d;
	t--;
	return c * (t * t * t * t * t + 1) + b;
};

// acceleration until halfway, then deceleration
kapi.tween.easeInOutQuint = function (t, b, c, d) {
	t /= d / 2;
	if (t < 1) {
		return c / 2 * t * t * t * t * t + b;
	}
	t -= 2;
	return c / 2 * (t * t * t * t * t + 2) + b;
};

// accelerating from zero velocity
kapi.tween.easeInSine = function (t, b, c, d) {
	return -c * Math.cos(t / d * (Math.PI / 2)) + c + b;
};

// decelerating to zero velocity
kapi.tween.easeOutSine = function (t, b, c, d) {
	return c * Math.sin(t / d * (Math.PI / 2)) + b;
};

// accelerating until halfway, then decelerating
kapi.tween.easeInOutSine = function (t, b, c, d) {
	return -c / 2 * (Math.cos(Math.PI * t / d) - 1) + b;
};

// accelerating from zero velocity
kapi.tween.easeInExpo = function (t, b, c, d) {
	return c * Math.pow(2, 10 * (t / d - 1)) + b;
};

// decelerating to zero velocity
kapi.tween.easeOutExpo = function (t, b, c, d) {
	return c * (-Math.pow(2, -10 * t / d) + 1) + b;
};

// accelerating until halfway, then decelerating
kapi.tween.easeInOutExpo = function (t, b, c, d) {
	t /= d / 2;
	if (t < 1) {
		return c / 2 * Math.pow(2, 10 * (t - 1)) + b;
	}
	t--;
	return c / 2 * (-Math.pow(2, -10 * t) + 2) + b;
};

// accelerating from zero velocity
kapi.tween.easeInCirc = function (t, b, c, d) {
	t /= d;
	return -c * (Math.sqrt(1 - t * t) - 1) + b;
};

// decelerating to zero velocity
kapi.tween.easeOutCirc = function (t, b, c, d) {
	t /= d;
	t--;
	return c * Math.sqrt(1 - t * t) + b;
};

// acceleration until halfway, then deceleration
kapi.tween.easeInOutCirc = function (t, b, c, d) {
	t /= d / 2;
	if (t < 1) {
		return -c / 2 * (Math.sqrt(1 - t * t) - 1) + b;
	}
	t -= 2;
	return c / 2 * (Math.sqrt(1 - t * t) + 1) + b;
}

demo_proj/demo_actors.js
/*global window:true, Image:true */

/** Kapi demo animation - actors
    by Jeremy Kahn - jeremyckahn@gmail.com
*/

(function () {
	if (!window._demoApp) {
		window._demoApp = {};
	}
	
	var app = window._demoApp,
		NUM_RAYS = 15,
		RAY_RADIUS = 320,
		RAY_BASE_WIDTH = 30;
	
	app.imageUrls = {
		
	};
	
	app.actors = {};
	app.actors.rays = function (ctx) {
		var i,
			color = '#fff',
			rayRot,
			adjustedAngle1,
			adjustedAngle2,
			adjustedAngle3;

		// A simple optimization.  If the opacity is `0`, then the rays are not visible,
		// and there is no reason to draw them at all.  Just return.
		if (this.alpha === 0) {
			return;
		}

		ctx.beginPath();
		ctx.globalAlpha = this.alpha;

		// Loop through all the rays and draw them at their respective angles
		for (i = 0; i < NUM_RAYS; i++) {
			ctx.moveTo(this.x, this.y);
			rayRot = window._demoApp.utils.degToRad( 360 * ( i / NUM_RAYS ) );
			
			adjustedAngle1 = this.rotate + rayRot - 90;
			adjustedAngle2 = this.rotate + rayRot;
			adjustedAngle3 = this.rotate + rayRot + 90;

			ctx.lineTo(
				this.x + (RAY_BASE_WIDTH / 2) * Math.sin( adjustedAngle1 ),
				this.y + (RAY_BASE_WIDTH / 2) * Math.cos( adjustedAngle1 ));

			ctx.lineTo(
				this.x + RAY_RADIUS * Math.sin( adjustedAngle2 ),
				this.y + RAY_RADIUS * Math.cos( adjustedAngle2 ));

			ctx.lineTo(
				this.x + (RAY_BASE_WIDTH / 2) * Math.sin( adjustedAngle3 ),
				this.y + (RAY_BASE_WIDTH / 2) * Math.cos( adjustedAngle3 ));
		}

		ctx.lineWidth = 1;
		ctx.fillStyle = ctx.strokeStyle = color;
		ctx.fill();
		ctx.stroke();
		ctx.closePath();
		ctx.globalAlpha = 1;
	};
	
	app.actors.img = function (src) {
		return {
			'setup': function setup () {
				this.img = new Image();
				// Call the `imageLoadComplete` function when the image is loaded.
				// Once the last image is loaded, the animation will `play()`.
				this.img.onload = window._demoApp.imageLoadComplete;
				this.img.src = src;
			},
			'draw': function draw (ctx) {
			  ctx.rotate(window._demoApp.utils.degToRad(this.rotate || 0));
				ctx.globalAlpha = this.alpha;
				ctx.drawImage(this.prototype.img, this.x, this.y, this.scaleX, this.scaleY);
				ctx.globalAlpha = 1;
				ctx.rotate(-window._demoApp.utils.degToRad(this.rotate || 0));
			} 
		};
	};
}());

ext/sequence/kapi.sequence.js
/*global kapi:true, console:true */

// Sequences should not end before they begins.

/**
 *`state` reference:
 *
 *  0 = Sequence hasn't started playing
 *  1 = Sequence has started playing
 *  2 = Sequence has played and been stopped
 */

(function (kapi) {
	var sequences;
	
	// Don't allow the sequence to start on keyframe 0, that will overwrite the first keyframe
	function validateKeyframeId (keyframeId) {
		if (keyframeId === 0 || keyframeId === '0' || keyframeId === '0s') {
			keyframeId = 1;
		}
		
		return keyframeId;
	}
	
	sequences = {};
	
	kapi.fn.sequence = {
		/**
		 * @param {String} sequenceName
		 * @param {Function} sequence
		 */
		_create: function create (sequenceName, sequence) {
			var master,
				exposedMasterData,
				actorSequenceName,
				actorTemplate,
				actorInst,
				puppetInst,
				puppetKeyframeIndex,
				masterPlayHandler,
				masterPauseHandler,
				masterStopHandler,
				masterLoopComplete;
				
			masterPlayHandler = function () {
				var masterFrame,
					puppetStartFrame;

				masterFrame = master._expose()._currentFrame;
				puppetStartFrame = master._expose()._actorstateIndex[actorSequenceName][1];

				if (masterFrame >= puppetStartFrame) {
					puppetInst.play();
					actorInst.data().state = 1;
				}
			};

			masterPauseHandler = function () {
				puppetInst.pause();
			};

			masterStopHandler = function () {
				puppetInst.stop();
			};
			
			// If the sequence is finished and the master's loop has completed,
			// reset the sequence's recorded state
			masterLoopComplete = function () {
				if (actorInst.data().state === 2) {
					actorInst.data().state = 0;
				}
			};
			
			actorTemplate = {
				setup: function () {
					
				},
				
				draw: function () {	

					// Once the starting keyframe has been passed, and the sequence actor has not begun,
					// it is time to `play()`
					if (actorInst.data().startingKeyframeId <= exposedMasterData._currentFrame + 1 && actorInst.data().state < 1) {
						puppetInst.iterate(1, function () {
							actorInst.data().state = 0;
						});
						actorInst.data().state = 1;
						
					// Once the ending keyframe has been passed, and the sequence actor has begun,
					// it is time to `stop()`
					} else if (actorInst.data().endingKeyframeId <= exposedMasterData._currentFrame + 1 && actorInst.data().state < 2) {
						puppetInst.stop();
						actorInst.data().state = 2;
					}
				},
				
				teardown: function () {
					master.unbind('onPlay', masterPlayHandler);
					master.unbind('onPause', masterPauseHandler);
					master.unbind('onStop', masterPauseHandler);
					master.unbind('loopComplete', masterLoopComplete);
				}
			};
				
			master = this;
			exposedMasterData = master._expose();
			actorSequenceName = '_sequence.' + sequenceName;
			master.bind('onPlay', masterPlayHandler);
			master.bind('onPause', masterPauseHandler);
			master.bind('onStop', masterStopHandler);
			master.bind('loopComplete', masterLoopComplete);
			puppetInst = master.puppetCreate(actorSequenceName, actorTemplate);
			sequence(puppetInst);
			
			actorInst = master.add(actorTemplate, {
				'name': actorSequenceName,
				'dummyProp': 0
			}).keyframe(0, {
				// Nothing!
			});
			
			actorInst.data({
				'state': 0
			});
			
			puppetKeyframeIndex = master._expose()._actorstateIndex[actorSequenceName];
			sequences[sequenceName] = actorInst;
		},
		
		/**
		 * @param {String} sequenceName
		 * @param {String|Number} keyframeId
		 */
		_startAt: function startAt (sequenceName, keyframeId) {
			var sequence;
			
			sequence = sequences[sequenceName];
			keyframeId = validateKeyframeId(keyframeId);
			
			// Make a dummy keyframe, so that the sequence registers a valid keyframe
			// on the master timeline
			sequence.keyframe(keyframeId, {
				'dummyProp': 1
			});
			
			sequence.data().startingKeyframeId = this._expose()._puppets[sequence.id].getRealKeyframe(keyframeId);
		},
		
		/**
		 * @param {String} sequenceName
		 * @param {String|Number} keyframeId
		 */
		_endAt: function endAt (sequenceName, keyframeId) {
			var sequence;
			
			sequence = sequences[sequenceName];
			keyframeId = validateKeyframeId(keyframeId);
			
			sequence.keyframe(keyframeId, {
				'dummyProp': -1
			});
			
			sequence.data().endingKeyframeId = this._expose()._puppets[sequence.id].getRealKeyframe(keyframeId);
		},
		
		/**
		 * @param {String} sequenceName
		 */
		_destroy: function destroy (sequenceName) {
			var sequence,
				puppet;
			
			sequence = sequences[sequenceName];
			puppet = sequence.kapi._expose()._puppets[sequence.id];
			this.removeActor(sequence.id);
			
			puppet.stop();
			delete sequence.kapi._expose()._puppets[sequence.id];
		}
	};
} (kapi));

demo_proj/demo.js
/*global kapi:true, window:true, document:true, Image:true */

/** Kapi demo animation
    by Jeremy Kahn - jeremyckahn@gmail.com

This is a demonstration of the type of animation that can be made with the Kapi, a keyframing API for the HTML 5 canvas.  Images are used under the Creative Commons License, and were obtained from: http://www.w3.org/html/logo/

For more information and to download Kapi, please visit: https://github.com/jeremyckahn/kapi
*/

(function () {
	var FRAME_RATE = 30,
		CANVAS_HEIGHT = 375,
		CANVAS_WIDTH = 500,
		BADGE_SIZE = 256,
		BADGE_OFFSET_TOP = 20,
		SUB_TECH_Y = 300,
		SUB_TECH_SIZE = 40,
		SUB_TECH_SPACING = 10,
		actors = {},
		subTechNameList = [],
		imageUrls = {
			'badge': 'img/HTML5_Badge.png',
			'3d': 'img/3D_Effects.png',
			'connect': 'img/Connectivity.png',
			'access': 'img/Device_Access.png',
			'multi': 'img/Multimedia.png',
			'offline': 'img/Offline_Storage.png',
			'perf': 'img/Performance.png',
			'sem': 'img/Semantics.png',
			'style': 'img/Styling.png'},
		numImages = 0,
		img,
		canvas,
		demo,
		actor,
		i;
		
	if (!window._demoApp) {
		window._demoApp = {};
	}
	
	for (img in imageUrls) {
		if (imageUrls.hasOwnProperty(img)) {
			numImages++;
		}
	}
		
	canvas = document.getElementsByTagName('canvas')[0];
	
	/////////////////
	// KAPI INSTANCE SETUP
	/////////////////
	
	window._demoApp.kapiInst = demo = kapi(canvas, {
		'fps': FRAME_RATE,
		styles: {
			'background': '#f2f2f2',
			'height': CANVAS_HEIGHT,
			'width': CANVAS_WIDTH
		}
	});
	
	// The animation will begin as soon as all of the images have loaded.
	window._demoApp.imageLoadComplete = function () {
		numImages--;
		if (numImages === 0) {
			window._demoApp.kapiInst.play();
		}
	};
	
	// Set a pause toggle for when the user clicks the canvas.
	canvas.addEventListener('click', function () {
		if (demo.isPlaying()) {
			demo.pause();
		}  else {
			demo.play();
		}
	}, true);
	
	/////////////////
	// ACTOR SETUP
	/////////////////
	
	actors.rays = demo.add(window._demoApp.actors.rays, {
		x: CANVAS_WIDTH / 2,
		y: 140,
		rotate: 0,
		alpha: 0
	});
	
	// Setup all the image actors and add them to the kapi instance
	for (actor in imageUrls) {
		if (imageUrls.hasOwnProperty(actor)) {
			
			actors[actor] = window._demoApp.actors.img(imageUrls[actor]);
			actors[actor] = demo.add(actors[actor], {
				'name': actor,
				x: 0,
				y: 0,
				scaleX: 0,
				scaleY: 0,
				alpha: 1,
				easing:'easeInOutQuint'
			});
			
			// All of the little black symbols (sub-techs) will be animated separately from the badge, 
			// this is a helpful way to keep track of them and loop through later
			if (actor !== 'badge') {
				subTechNameList.push(actor);
			}
		}
	}
	
	/////////////////
	// KEYFRAME SETUP
	/////////////////
	
	// Keyframes for the light rays.
	actors.rays
		.keyframe(0, {
			
		})
		.keyframe('1s', {})
		.keyframe('2s', {
			rotate: window._demoApp.utils.revsPerSecond(1),
			alpha: 1
		})
		.keyframe('6s', {
			rotate: window._demoApp.utils.revsPerSecond(4)
		}).keyframe('6.5s', {
			alpha: 0,
			rotate: window._demoApp.utils.revsPerSecond(0.5)
		});
	
	// Keyframes for the big HTML5 badge image.
	actors.badge
		.keyframe(0, {
			scaleX: 1,
			scaleY: 1,
			alpha: 0,
			rotate: 45
		})
		.keyframe('1.75s', {
			scaleX: BADGE_SIZE,
			scaleY: BADGE_SIZE,
			x: (CANVAS_WIDTH / 2) - (BADGE_SIZE / 2),
			y: BADGE_OFFSET_TOP,
			alpha: 1,
			rotate: 0
		})
		.keyframe('4s', {})
		.keyframe('5s', {
			scaleX: BADGE_SIZE + 70,
			scaleY: BADGE_SIZE + 70,
			x: '-=35'
		})
		.keyframe('6s', {})
		.keyframe('7s', {
			alpha: 0
		});
		
	// Helper functions for sub-tech actor positioning.  Seems a bit messier, but JSLint yells at you
	// if you define the functions inside of the for loop.  These are used in the loop below.
	function getXofOffscreenSubTech () {
		return -this.scaleX;
	}
	
	function getXofOnScreenSubTech (index) {
		return function () {
			return ((this.scaleX + SUB_TECH_SPACING) * index) + 50;
		};
	}
	
	// Loop through all of the "sub-tech" image actors and keyframe them.
	for (i = 0; i < subTechNameList.length; i++) {
		// We have the name of the actor, so grab it out of the `actors` Object
		actor = actors[subTechNameList[i]];
		
		actor
			.keyframe(0, {
				y: SUB_TECH_Y,
				x: getXofOffscreenSubTech,
				alpha: 0,
				scaleX: SUB_TECH_SIZE,
				scaleY: SUB_TECH_SIZE
			})
			.keyframe('1s', {
				x: getXofOnScreenSubTech(i),
				alpha: 1
			})
			.keyframe('4s', {})
			.keyframe('4.5s', {
				alpha: 0
			});
	}
	
}());